src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJFS.java

```diff
@@ -181,22 +213,50 @@ FileMeta getMetadata( final File target, final LocalLockOwner owner ) throws IOE
             return lockManager.reentrantSynchronous( path, ( opLock ) -> {
 
                 TransactionManager transactionManager = metadataCache.getAdvancedCache().getTransactionManager();
-                transactionManager.begin();
+
+                try
+                {
+                    transactionManager.begin();
+                }
+                catch ( NotSupportedException | SystemException e )
+                {
+                    throw new PartylineException(
+                                    "Problem starting transaction obtaining metadata with path: " + path );
+                }
+
                 FileMeta meta = metadataCache.computeIfAbsent( path, ( p ) -> new FileMeta( p, target.isDirectory() ) );
+
                 // Check the current lock level
-                // TODO: Negotiate the lock level
                 LockLevel currentLockLevel = meta.getLockLevel( this.nodeKey );
-
-                meta.addLock( this.nodeKey, owner.getLockLevel() );
-                metadataCache.put( path, meta );
-                transactionManager.commit();
+                if ( currentLockLevel == null || currentLockLevel != owner.getLockLevel() )
+                {
+                    meta.addLock( this.nodeKey, owner.getLockLevel() );
+                    // Only update the cache if the lock level changed
+                    metadataCache.put( path, meta );
+                }
+                try
```

Comment:
This should all happen in one try/catch section, I believe. I'm not sure it's useful to separate them, and if you do you lose the ability to use try/catch/finally to force some kind of action to be taken WRT the transaction. I recommend something of the form:

```
TransactionManager transactionManager = ...;
try
{
    transactionManager.begin();

    // do some stuff
}
catch (SomeException e)
{
    transactionManager.rollback(); // catch + log any exceptions related to this, to the ERROR log level
    throw new PartylineException( "Failed to do something something. Rolled back transaction", e);
}
finally
{
    transactionManager.commit(); // catch + log exceptions here
}
```